### PR TITLE
Add BLS_BLS12381 signature algorithm for cadence

### PR DESCRIPTION
**Pull request by SupunS**
## Description

Work towards #582

______

For contributor use:

- [ ] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/crypto_test.go [hunk 1]
```diff
@@ -187,3 +189,43 @@ func TestRuntimeCrypto_hash(t *testing.T) {
 
 	assert.True(t, called)
 }
+
+func TestSignatureAlgorithms(t *testing.T) {
```
Comment (turbolent) [inline runtime/crypto_test.go:193]: 👌

**Review by turbolent — APPROVED** [review]

File: docs/language/crypto.md [hunk 2]
```diff
@@ -33,6 +33,9 @@ pub enum SignatureAlgorithm: UInt8 {
 
     /// ECDSA_Secp256k1 is Elliptic Curve Digital Signature Algorithm (ECDSA) on the secp256k1 curve.
     pub case ECDSA_Secp256k1 = 2
+
+    /// BLS_BLS12381 is BLS signature scheme on the BLS12381 curve.
```
Comment (tarakby) [inline docs/language/crypto.md:37]: Just to use the original [name](https://electriccoin.co/blog/new-snark-curve/) of the curve. 
```suggestion
    /// BLS_BLS12381 is BLS signature scheme on the BLS12-381 curve.
```

Comment (tarakby) [reply docs/language/crypto.md:37]: You might be asking why I used `BLS_BLS12381` and not `BLS_BLS12_381`:
I used the underscore in `ECDSA` to separate the scheme name and the parameter name (here the curve), so I didn't want to add an extra underscore in the curve name `BLS12_381`. I'm open to better suggestions!

File: runtime/crypto_test.go [hunk 3]
```diff
@@ -187,3 +189,43 @@ func TestRuntimeCrypto_hash(t *testing.T) {
...
+
+	t.Parallel()
+
+	runtime := NewInterpreterRuntime()
+	runtimeInterface := &testRuntimeInterface{}
+
+	testHashAlgorithm := func(algo sema.CryptoAlgorithm) {
```
Comment (tarakby) [inline runtime/crypto_test.go:200]: typo?
```suggestion
	testSignatureAlgorithm := func(algo sema.CryptoAlgorithm) {
```

File: runtime/sema/crypto_algorithm_types.go [hunk 4]
```diff
@@ -196,6 +204,10 @@ const SignatureAlgorithmDocStringECDSA_Secp256k1 = `
 ECDSA_Secp256k1 is Elliptic Curve Digital Signature Algorithm (ECDSA) on the secp256k1 curve
 `
 
+const SignatureAlgorithmDocStringBLS_BLS12381 = `
+BLS_BLS12381 is BLS signature scheme on the BLS12381 curve
```
Comment (tarakby) [inline runtime/sema/crypto_algorithm_types.go:208]:
```suggestion
BLS_BLS12381 is BLS signature scheme on the BLS12-381 curve
```

File: runtime/crypto_test.go [hunk 5]
```diff
@@ -187,3 +189,43 @@ func TestRuntimeCrypto_hash(t *testing.T) {
...
+
+		require.IsType(t, cadence.Enum{}, value)
+		enumValue := value.(cadence.Enum)
+
+		require.Len(t, enumValue.Fields, 1)
+		assert.Equal(t, cadence.NewUInt8(algo.RawValue()), enumValue.Fields[0])
+	}
+
+	for _, algo := range sema.SignatureAlgorithms {
+		testHashAlgorithm(algo)
```
Comment (tarakby) [inline runtime/crypto_test.go:229]:
```suggestion
		testSignatureAlgorithm(algo)
```

**Review by tarakby — APPROVED** [review]
👏

File: docs/language/crypto.md (hunk 2 above)
Comment (turbolent) [reply docs/language/crypto.md:37]: Oh, thanks for the explanation! In that case I'd recommend `BLS_BLS12_381`, because it's more readable IMHO

Comment (tarakby) [reply docs/language/crypto.md:37]: That works for me!

File: runtime/crypto_test.go (hunk 5 above)
Comment (SupunS) [reply runtime/crypto_test.go:229]: oh, I copied this from the other, and forgot to update!

File: docs/language/crypto.md (hunk 2 above)
Comment (SupunS) [reply docs/language/crypto.md:37]: In `KMAC128_BLS_BLS12381` hashing algo, we use `BLS_BLS12381`. Should we use the same pattern here as well (or use `BLS_BLS12_381` there - but that would make the name really long)?

I'm not sure whether they both refer to the same thing though.

Comment (tarakby) [reply docs/language/crypto.md]: They do refer to the same thing, it's better to be consistent and have `BLS_12_381` for both 🤷‍♂️

Comment (SupunS) [reply docs/language/crypto.md]: Updated to use `BLS_BLS12_381` format for the signature algo in 1e25c63.
I will send a separate PR updating the hash algorithm to use the same.

**Review by turbolent — APPROVED** [review]
Looks good! 👍

**Pull request merged**